aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move

```diff
@@ -0,0 +1,529 @@
+///
+/// Currently, one implementation is provided, backed by a single sorted vector.
+///
+/// That means that keys can be found within O(log N) time.
+/// Adds and removals take O(N) time, but the constant factor is small,
+/// as it does only O(log N) comparisons, and does efficient mem-copy with vector operations.
+///
+/// Additionally, it provides a way to lookup and iterate over sorted keys, making range query
+/// take O(log N + R) time (where R is number of elements in the range).
+///
+/// Most methods operate with OrderedMap being `self`.
+/// All methods that start with iter_*, operate with Iterator being `self`.
+///
+/// Uses cmp::compare for ordering, which compares primitive types natively, and uses common
+/// lexicographical sorting for complex types.
+module aptos_std::ordered_map {
+    // friend aptos_std::big_ordered_map;
+
+    use std::vector;
+
+    use std::option::{Self, Option};
+    use std::cmp;
+    use std::error;
+    use std::mem;
+
+    /// Map key already exists
+    const EKEY_ALREADY_EXISTS: u64 = 1;
+    /// Map key is not found
+    const EKEY_NOT_FOUND: u64 = 2;
+    // Trying to do an operation on an Iterator that would go out of bounds
+    const EITER_OUT_OF_BOUNDS: u64 = 3;
+    /// New key used in replace_key_inplace doesn't respect the order
+    const ENEW_KEY_NOT_IN_ORDER: u64 = 4;
+
+    /// Individual entry holding (key, value) pair
+    struct Entry<K, V> has drop, copy, store {
+        key: K,
+        value: V,
+    }
+
+    /// The OrderedMap datastructure.
+    enum OrderedMap<K, V> has drop, copy, store {
+        /// sorted-vector based implementation of OrderedMap
+        SortedVectorMap {
+            /// List of entries, sorted by key.
+            entries: vector<Entry<K, V>>,
+        }
+    }
+
+    /// An iterator pointing to a position between two elements in the
+    enum Iterator has copy, drop {
+        // TODO: see if it is cleaner/faster to have end-iter be Position { index = len() }
+        // and not have a separate variant for it.
+        End,
+        Position {
+            /// The index of the iterator pointing to.
+            index: u64,
+        },
+    }
+
+    /// Creates a new empty OrderedMap, using default (SortedVectorMap) implementation.
+    public fun new<K, V>(): OrderedMap<K, V> {
+        OrderedMap::SortedVectorMap {
+            entries: vector::empty(),
+        }
+    }
+
+    /// Number of elements in the map.
+    public fun length<K, V>(self: &OrderedMap<K, V>): u64 {
+        self.entries.length()
+    }
+
+    /// Whether map is empty.
+    public fun is_empty<K, V>(self: &OrderedMap<K, V>): bool {
+        self.entries.is_empty()
+    }
+
+    /// Add a key/value pair to the map.
+    /// Aborts with EKEY_ALREADY_EXISTS if key already exist.
+    public fun add<K, V>(self: &mut OrderedMap<K, V>, key: K, value: V) {
+        let len = self.entries.length();
+        let index = binary_search(&key, &self.entries, 0, len);
```

Comment:
we should introduce iterator to vector too so you don't need to know the length anymore for assertion.

Reply:
this seems fine to me